README

This Module runs with the Tryton application platform.

Installing
----------

See INSTALL

Support
-------

If you encounter any problems with this module, please don't hesitate to ask
questions on the NaN·tic bug tracker, mailing list,
wiki or IRC channel:

  http://doc.tryton-erp.es/
  http://bitbucket.org/nantic/aeat_349
  http://groups.tryton.org/
  http://wiki.tryton.org/
  irc://irc.freenode.net/tryton

License
-------

See LICENSE

Copyright
---------

See COPYRIGHT


For more information please visit the NaN·tic web site:

  http://www.nan-tic.com/


CHANGELOG
Version 5.5.0 - 2019-11-14
Version 5.4.0 - 2019-11-14
Version 5.2.0 - 2019-05-07
Version 5.3.0 - 2019-05-06
Version 4.2.0 - 2016-11-28
* Bug fixes (see mercurial logs for details)

Version 4.0.0 - 2016-05-03
Version 3.4.0 - 2014-11-03

Version 3.2.0 - 2014-05-22

Version 3.0.0 - 2013-10-25
* Initial release


tests/__init__.py
# The COPYRIGHT file at the top level of this repository contains the full
# copyright notices and license terms.
try:
    from trytond.modules.aeat_349.tests.test_aeat_349 import suite
except ImportError:
    from .test_aeat_349 import suite

__all__ = ['suite']


tests/test_aeat_349.py
# This file is part of the aeat_349 module for Tryton.
# The COPYRIGHT file at the top level of this repository contains the full
# copyright notices and license terms.
import unittest
import doctest
import trytond.tests.test_tryton
from trytond.tests.test_tryton import ModuleTestCase
from trytond.tests.test_tryton import doctest_teardown
from trytond.tests.test_tryton import doctest_checker


class Aeat349TestCase(ModuleTestCase):
    'Test Aeat 349 module'
    module = 'aeat_349'


def suite():
    suite = trytond.tests.test_tryton.suite()
    suite.addTests(unittest.TestLoader().loadTestsFromTestCase(
        Aeat349TestCase))
    suite.addTests(doctest.DocFileSuite('scenario_aeat349.rst',
            tearDown=doctest_teardown, encoding='utf-8',
            checker=doctest_checker,
            optionflags=doctest.REPORT_ONLY_FIRST_FAILURE))
    suite.addTests(doctest.DocFileSuite(
            'scenario_aeat349_alternate_currency.rst',
            tearDown=doctest_teardown, encoding='utf-8',
            checker=doctest_checker,
            optionflags=doctest.REPORT_ONLY_FIRST_FAILURE))
    return suite


invoice.py
from trytond.model import ModelSQL, ModelView, fields, Unique
from trytond.wizard import Wizard, StateView, StateTransition, Button
from trytond.pool import Pool, PoolMeta
from trytond.pyson import Eval
from trytond.transaction import Transaction
from sql.operators import In
from sql import Literal
from .aeat import OPERATION_KEY

__all__ = ['Type', 'TypeTax', 'TypeTaxTemplate', 'Record', 'TaxTemplate',
    'Tax', 'Invoice', 'InvoiceLine', 'Recalculate349RecordStart',
    'Recalculate349RecordEnd', 'Recalculate349Record', 'Reasign349RecordStart',
    'Reasign349RecordEnd', 'Reasign349Record']


class Type(ModelSQL, ModelView):
    """
    AEAT 349 Type

    Keys types for AEAT 349 Report
    """
    __name__ = 'aeat.349.type'

    operation_key = fields.Selection(OPERATION_KEY, 'Operation key',
        required=True)

    @classmethod
    def __setup__(cls):
        super(Type, cls).__setup__()
        t = cls.__table__()
        cls._sql_constraints += [
            ('operation_key_uniq', Unique(t, t.operation_key),
                'Operation key must be unique.')
            ]

    def get_rec_name(self, name):
        opts = self.fields_get('operation_key')['operation_key']['selection']
        for key, value in opts:
            if self.operation_key == key:
                return value
        return self.operation_key


class TypeTax(ModelSQL):
    """
    AEAT 349 Type-Tax Relation
    """
    __name__ = 'aeat.349.type-account.tax'

    aeat_349_type = fields.Many2One('aeat.349.type', 'Operation Key',
        ondelete='CASCADE', select=True, required=True)
    tax = fields.Many2One('account.tax', 'Tax', ondelete='CASCADE',
        select=True, required=True)


class TypeTaxTemplate(ModelSQL):
    """
    AEAT 349 Type-Tax Template Relation
    """
    __name__ = 'aeat.349.type-account.tax.template'

    aeat_349_type = fields.Many2One('aeat.349.type', 'Operation Key',
        ondelete='CASCADE', select=True, required=True)
    tax = fields.Many2One('account.tax.template', 'Tax Template',
        ondelete='CASCADE', select=True, required=True)

    @classmethod
    def __register__(cls, module_name):
        pool = Pool()
        ModelData = pool.get('ir.model.data')
        Module = pool.get('ir.module')
        cursor = Transaction().connection.cursor()
        module_table = Module.__table__()
        sql_table = ModelData.__table__()
        # Meld aeat_349_es into aeat_349
        cursor.execute(*module_table.update(
                columns=[module_table.state],
                values=[Literal('uninstalled')],
                where=module_table.name == Literal('aeat_349_es')
                ))
        cursor.execute(*sql_table.update(
                columns=[sql_table.module],
                values=[module_name],
                where=sql_table.module == Literal('aeat_349_es')))
        super(TypeTaxTemplate, cls).__register__(module_name)


class Record(ModelSQL, ModelView):
    """
    AEAT 349 Record

    Calculated on invoice creation to generate temporal
    data for reports. Aggregated on aeat349 calculation.
    """
    __name__ = 'aeat.349.record'

    company = fields.Many2One('company.company', 'Company', required=True,
        readonly=True)
    fiscalyear = fields.Many2One('account.fiscalyear', 'Fiscal Year',
        required=True, readonly=True)
    month = fields.Integer('Month', readonly=True)
    party_vat = fields.Char('VAT', size=17, readonly=True)
    party_name = fields.Char('Party Name', size=40, readonly=True)
    operation_key = fields.Selection(OPERATION_KEY, 'Operation key',
        required=True, readonly=True)
    base = fields.Numeric('Base Operation Amount', digits=(16, 2),
        readonly=True)
    invoice = fields.Many2One('account.invoice', 'Invoice', readonly=True)
    operation = fields.Many2One('aeat.349.report.operation', 'Operation',
        readonly=True)


class TaxTemplate(ModelSQL, ModelView):
    'Account Tax Template'
    __name__ = 'account.tax.template'
    aeat349_operation_keys = fields.Many2Many(
        'aeat.349.type-account.tax.template', 'tax', 'aeat_349_type',
        'Available Operation Keys')
    aeat349_default_out_operation_key = fields.Many2One('aeat.349.type',
        'Default Out Operation Key',
        domain=[('id', 'in', Eval('aeat349_operation_keys', []))],
        depends=['aeat349_operation_keys'])
    aeat349_default_in_operation_key = fields.Many2One('aeat.349.type',
        'Default In Operation Key',
        domain=[('id', 'in', Eval('aeat349_operation_keys', []))],
        depends=['aeat349_operation_keys'])

    def _get_tax_value(self, tax=None):
        res = super(TaxTemplate, self)._get_tax_value(tax)

        old_ids = set()
        new_ids = set()
        if tax and len(tax.aeat349_operation_keys) > 0:
            old_ids = set([c.id for c in tax.aeat349_operation_keys])
        if len(self.aeat349_operation_keys) > 0:
            new_ids = set([c.id for c in self.aeat349_operation_keys])
            for direction in ('in', 'out'):
                field = "aeat349_default_%s_operation_key" % (direction)
                if not tax or getattr(tax, field) != getattr(self, field):
                    value = getattr(self, field)
                    if value and value.id in new_ids:
                        res[field] = value.id
                    else:
                        res[field] = None
        else:
            if tax and tax.aeat349_default_in_operation_key:
                res['aeat349_default_in_operation_key'] = None
            if tax and tax.aeat349_default_out_operation_key:
                res['aeat349_default_out_operation_key'] = None
        if old_ids or new_ids:
            key = 'aeat349_operation_keys'
            res[key] = []
            to_remove = old_ids - new_ids
            if to_remove:
                res[key].append(['remove', list(to_remove)])
            to_add = new_ids - old_ids
            if to_add:
                res[key].append(['add', list(to_add)])
            if not res[key]:
                del res[key]
        return res


class Tax(metaclass=PoolMeta):
    __name__ = 'account.tax'

    aeat349_operation_keys = fields.Many2Many('aeat.349.type-account.tax',
        'tax', 'aeat_349_type', 'Available Operation Keys')
    aeat349_default_out_operation_key = fields.Many2One('aeat.349.type',
        'Default Out Operation Key',
        domain=[('id', 'in', Eval('aeat349_operation_keys', []))],
        depends=['aeat349_operation_keys'])
    aeat349_default_in_operation_key = fields.Many2One('aeat.349.type',
        'Default In Operation Key',
        domain=[('id', 'in', Eval('aeat349_operation_keys', []))],
        depends=['aeat349_operation_keys'])


STATES = {
    'invisible': Eval('type') != 'line',
    }
DEPENDS = ['type']


class InvoiceLine(metaclass=PoolMeta):
    __name__ = 'account.invoice.line'
    aeat349_available_keys = fields.Function(fields.Many2Many('aeat.349.type',
        None, None, 'AEAT 349 Available Keys',
        states=STATES, depends=DEPENDS + ['taxes', 'product']),
        'on_change_with_aeat349_available_keys')
    aeat349_operation_key = fields.Many2One('aeat.349.type',
        'AEAT 349 Operation Key',
        states=STATES, depends=DEPENDS + ['aeat349_available_keys', 'taxes',
            'invoice_type', 'product'],
        domain=[('id', 'in', Eval('aeat349_available_keys', []))],)

    @fields.depends('invoice', 'taxes')
    def on_change_product(self):
        Taxes = Pool().get('account.tax')

        super(InvoiceLine, self).on_change_product()
        type_ = None
        if self.invoice and self.invoice.type:
            type_ = self.invoice.type
        elif self.invoice_type:
            type_ = self.invoice_type
        self.aeat349_operation_key = None
        if type_ and self.taxes:
            with Transaction().set_user(0):
                taxes = Taxes.browse(self.taxes)
            self.aeat349_operation_key = self.get_aeat349_operation_key(
                        type_, taxes)

    @fields.depends('taxes', 'product')
    def on_change_with_aeat349_available_keys(self, name=None):
        keys = []
        for tax in self.taxes:
            keys.extend([k.id for k in tax.aeat349_operation_keys])
        return list(set(keys))

    @fields.depends('taxes', 'invoice_type', 'aeat349_operation_key',
        '_parent_invoice.type', 'product', 'invoice')
    def on_change_with_aeat349_operation_key(self):
        if self.aeat349_operation_key:
            return self.aeat349_operation_key.id

        if self.invoice and self.invoice.type:
            type_ = self.invoice.type
        elif self.invoice_type:
            type_ = self.invoice_type
        else:
            return

        return self.get_aeat349_operation_key(type_, self.taxes)

    @classmethod
    def get_aeat349_operation_key(cls, invoice_type, taxes):
        type_ = 'in' if invoice_type == 'in' else 'out'
        for tax in taxes:
            name = 'aeat349_default_%s_operation_key' % type_
            value = getattr(tax, name)
            if value:
                return value.id

    @classmethod
    def create(cls, vlist):
        Invoice = Pool().get('account.invoice')
        Taxes = Pool().get('account.tax')
        for vals in vlist:
            if vals.get('type', 'line') != 'line':
                continue
            if not vals.get('aeat349_operation_key') and vals.get('taxes'):
                invoice_type = vals.get('invoice_type')
                if not invoice_type and vals.get('invoice'):
                    invoice = Invoice(vals.get('invoice'))
                    invoice_type = invoice.type
                taxes_ids = []
                for key, value in vals.get('taxes'):
                    if key == 'add':
                        taxes_ids.extend(value)
                with Transaction().set_user(0):
                    taxes = Taxes.browse(taxes_ids)
                vals['aeat349_operation_key'] = cls.get_aeat349_operation_key(
                    invoice_type, taxes)
        return super(InvoiceLine, cls).create(vlist)


class Invoice(metaclass=PoolMeta):
    __name__ = 'account.invoice'

    @classmethod
    def create_aeat349_records(cls, invoices):
        pool = Pool()
        Record = pool.get('aeat.349.record')
        Currency = pool.get('currency.currency')
        to_create = {}
        for invoice in invoices:
            if not invoice.move or invoice.state == 'cancel':
                continue
            for line in invoice.lines:
                if line.type != 'line':
                    continue
                if line.aeat349_operation_key:
                    operation_key = line.aeat349_operation_key.operation_key
                    key = "%d-%s" % (invoice.id, operation_key)
                    amount = line.amount
                    if invoice.currency != invoice.company.currency:
                        with Transaction().set_context(
                                date=invoice.currency_date):
                            amount = Currency.compute(
                                invoice.currency, amount,
                                invoice.company.currency)
                    if key in to_create:
                        to_create[key]['base'] += amount
                    else:
                        month = (invoice.accounting_date.month
                             if invoice.accounting_date
                             else invoice.invoice_date.month)
                        to_create[key] = {
                            'company': invoice.company.id,
                            'fiscalyear': invoice.move.period.fiscalyear,
                            'month': month,
                            'party_name': invoice.party.name[:40],
                            'party_vat': (invoice.party.tax_identifier.code
                                if invoice.party.tax_identifier else ''),
                            'base': amount,
                            'operation_key': operation_key,
                            'invoice': invoice.id,
                            }

        with Transaction().set_user(0, set_context=True):
            Record.delete(Record.search([('invoice', 'in',
                            [i.id for i in invoices])]))
            Record.create(list(to_create.values()))

    @classmethod
    def draft(cls, invoices):
        pool = Pool()
        Record = pool.get('aeat.349.record')
        super(Invoice, cls).draft(invoices)
        with Transaction().set_user(0, set_context=True):
            Record.delete(Record.search([('invoice', 'in',
                            [i.id for i in invoices])]))

    @classmethod
    def post(cls, invoices):
        super(Invoice, cls).post(invoices)
        cls.create_aeat349_records(invoices)

    @classmethod
    def cancel(cls, invoices):
        pool = Pool()
        Record = pool.get('aeat.349.record')
        super(Invoice, cls).cancel(invoices)
        with Transaction().set_user(0, set_context=True):
            Record.delete(Record.search([('invoice', 'in',
                            [i.id for i in invoices])]))


class Recalculate349RecordStart(ModelView):
    """
    Recalculate AEAT 349 Records Start
    """
    __name__ = "aeat.349.recalculate.records.start"


class Recalculate349RecordEnd(ModelView):
    """
    Recalculate AEAT 349 Records End
    """
    __name__ = "aeat.349.recalculate.records.end"


class Recalculate349Record(Wizard):
    """
    Recalculate AEAT 349 Records
    """
    __name__ = "aeat.349.recalculate.records"
    start = StateView('aeat.349.recalculate.records.start',
        'aeat_349.aeat_349_recalculate_start_view', [
            Button('Cancel', 'end', 'tryton-cancel'),
            Button('Calculate', 'calculate', 'tryton-ok', default=True),
            ])
    calculate = StateTransition()
    done = StateView('aeat.349.recalculate.records.end',
        'aeat_349.aeat_349_recalculate_end_view', [
            Button('Ok', 'end', 'tryton-ok', default=True),
            ])

    def transition_calculate(self):
        Invoice = Pool().get('account.invoice')
        invoices = Invoice.browse(Transaction().context['active_ids'])
        Invoice.create_aeat349_records(invoices)
        return 'done'


class Reasign349RecordStart(ModelView):
    """
    Reasign AEAT 349 Records Start
    """
    __name__ = "aeat.349.reasign.records.start"

    aeat_349_type = fields.Many2One('aeat.349.type', 'Operation Key',
        required=True)


class Reasign349RecordEnd(ModelView):
    """
    Reasign AEAT 349 Records End
    """
    __name__ = "aeat.349.reasign.records.end"


class Reasign349Record(Wizard):
    """
    Reasign AEAT 349 Records
    """
    __name__ = "aeat.349.reasign.records"
    start = StateView('aeat.349.reasign.records.start',
        'aeat_349.aeat_349_reasign_start_view', [
            Button('Cancel', 'end', 'tryton-cancel'),
            Button('Reasign', 'reasign', 'tryton-ok', default=True),
            ])
    reasign = StateTransition()
    done = StateView('aeat.349.reasign.records.end',
        'aeat_349.aeat_349_reasign_end_view', [
            Button('Ok', 'end', 'tryton-ok', default=True),
            ])

    def transition_reasign(self):
        Invoice = Pool().get('account.invoice')
        Line = Pool().get('account.invoice.line')
        cursor = Transaction().connection.cursor()
        invoices = Invoice.browse(Transaction().context['active_ids'])

        value = self.start.aeat_349_type
        lines = []
        invoice_ids = set()
        for invoice in invoices:
            for line in invoice.lines:
                if value in line.aeat349_available_keys:
                    lines.append(line.id)
                    invoice_ids.add(invoice.id)

        if not invoice_ids:
            return 'done'

        line = Line.__table__()
        # Update to allow to modify key for posted invoices
        cursor.execute(*line.update(columns=[line.aeat349_operation_key],
                values=[value.id], where=In(line.id, lines)))

        invoices = Invoice.browse(list(invoices))
        Invoice.create_aeat349_records(invoices)

        return 'done'


__init__.py
# The COPYRIGHT file at the top level of this repository contains the full
# copyright notices and license terms.
from trytond.pool import Pool
from . import aeat
from . import invoice


def register():
    Pool.register(
        aeat.Report,
        aeat.Operation,
        aeat.Ammendment,
        invoice.Type,
        invoice.TypeTaxTemplate,
        invoice.TypeTax,
        invoice.Record,
        invoice.TaxTemplate,
        invoice.Tax,
        invoice.Invoice,
        invoice.InvoiceLine,
        invoice.Recalculate349RecordStart,
        invoice.Recalculate349RecordEnd,
        invoice.Reasign349RecordStart,
        invoice.Reasign349RecordEnd,
        module='aeat_349', type_='model')
    Pool.register(
        invoice.Recalculate349Record,
        invoice.Reasign349Record,
        module='aeat_349', type_='wizard')


aeat.py
# -*- coding: utf-8 -*-
import datetime
import itertools
from decimal import Decimal
import unicodedata
import sys

from retrofix import aeat349
from retrofix.record import Record, write as retrofix_write

from trytond.model import Workflow, ModelSQL, ModelView, fields
from trytond.pool import Pool
from trytond.pyson import Eval
from trytond.i18n import gettext
from trytond.exceptions import UserError
from trytond.transaction import Transaction

__all__ = ['Report', 'Operation', 'Ammendment']

PERIOD = [
    ('1T', 'First quarter'),
    ('2T', 'Second quarter'),
    ('3T', 'Third quarter'),
    ('4T', 'Fourth quarter'),
    ('01', 'January'),
    ('02', 'February'),
    ('03', 'March'),
    ('04', 'April'),
    ('05', 'May'),
    ('06', 'June'),
    ('07', 'July'),
    ('08', 'August'),
    ('09', 'September'),
    ('10', 'October'),
    ('11', 'November'),
    ('12', 'December'),
    ]

OPERATION_KEY = [
    ('E', 'E - Intra-Community supplies'),
    ('A', 'A - Intra-Community acquisition'),
    ('T', 'T - Triangular operations'),
    ('S', 'S - Intra-Community services'),
    ('I', 'I - Intra-Community services acquisitions'),
    ('M', 'M - Intra-Community supplies without taxes'),
    ('H', 'H - Intra-Community supplies without taxes delivered '
        'by legal representative'),
    ]

_ZERO = Decimal('0.0')


def remove_accents(unicode_string):
    str_ = str if sys.version_info < (3, 0) else bytes
    unicode_ = str if sys.version_info < (3, 0) else str
    if isinstance(unicode_string, str_):
        unicode_string_bak = unicode_string
        try:
            unicode_string = unicode_string_bak.decode('iso-8859-1')
        except UnicodeDecodeError:
            try:
                unicode_string = unicode_string_bak.decode('utf-8')
            except UnicodeDecodeError:
                return unicode_string_bak

    if not isinstance(unicode_string, unicode_):
        return unicode_string

    unicode_string_nfd = ''.join(
        (c for c in unicodedata.normalize('NFD', unicode_string)
            if (unicodedata.category(c) != 'Mn')
            ))
    # It converts nfd to nfc to allow unicode.decode()
    return unicodedata.normalize('NFC', unicode_string_nfd)


class Report(Workflow, ModelSQL, ModelView):
    """
    AEAT 349 Report
    """
    __name__ = "aeat.349.report"

    company = fields.Many2One('company.company', 'Company', required=True,
        states={
            'readonly': Eval('state') == 'done',
            }, depends=['state'])
    currency = fields.Function(fields.Many2One('currency.currency',
        'Currency'), 'get_currency')
    previous_number = fields.Char('Previous Declaration Number', size=13,
        states={
            'readonly': Eval('state') == 'done',
            'invisible': Eval('type') == 'N',
            'required': Eval('type') != 'N',
            }, depends=['state', 'type'])
    representative_vat = fields.Char('L.R. VAT number', size=9,
        help='Legal Representative VAT number.', states={
            'readonly': Eval('state') == 'done',
            }, depends=['state'])
    fiscalyear = fields.Many2One('account.fiscalyear', 'Fiscal Year',
        required=True, states={
            'readonly': Eval('state') == 'done',
            }, depends=['state'])
    fiscalyear_code = fields.Integer('Fiscal Year Code', required=True,
        depends=['fiscalyear'])
    company_vat = fields.Char('VAT number', size=9, states={
            'required': True,
            'readonly': Eval('state') == 'done',
            }, depends=['state', 'company'])
    type = fields.Selection([
            ('N', 'Normal'),
            ('C', 'Complementary'),
            ('S', 'Substitutive')
            ], 'Statement Type', required=True, states={
                'readonly': Eval('state') == 'done',
            }, depends=['state'])
    support_type = fields.Selection([
            ('C', 'DVD'),
            ('T', 'Telematics'),
            ], 'Support Type', required=True, states={
                'readonly': Eval('state') == 'done',
            }, depends=['state'])
    calculation_date = fields.DateTime("Calculation Date", readonly=True)
    state = fields.Selection([
            ('draft', 'Draft'),
            ('calculated', 'Calculated'),
            ('done', 'Done'),
            ('cancelled', 'Cancelled')
            ], 'State', readonly=True)
    period = fields.Selection(PERIOD, 'Period', sort=False, required=True)
    contact_name = fields.Char('Full Name', size=40,
        help='Must have name and surname.', states={
            'required': True,
            'readonly': Eval('state') == 'confirmed',
            }, depends=['state'])
    contact_phone = fields.Char('Phone', size=9, states={
            'required': True,
            'readonly': Eval('state') == 'confirmed',
            }, depends=['state'])
    operations = fields.One2Many('aeat.349.report.operation', 'report',
        'Operations')
    operation_amount = fields.Function(fields.Numeric(
            'Operation Amount', digits=(16, 2)), 'get_totals')
    ammendments = fields.One2Many('aeat.349.report.ammendment', 'report',
        'Ammendments')
    ammendment_amount = fields.Function(fields.Numeric(
            'Ammendment Amount', digits=(16, 2)), 'get_totals')
    file_ = fields.Binary('File', filename='filename', states={
            'invisible': Eval('state') != 'done',
            })
    filename = fields.Function(fields.Char("File Name"),
        'get_filename')

    @classmethod
    def __setup__(cls):
        super(Report, cls).__setup__()
        cls._buttons.update({
                'draft': {
                    'invisible': ~Eval('state').in_(['calculated',
                            'cancelled']),
                    },
                'calculate': {
                    'invisible': ~Eval('state').in_(['draft']),
                    },
                'process': {
                    'invisible': ~Eval('state').in_(['calculated']),
                    },
                'cancel': {
                    'invisible': Eval('state').in_(['cancelled']),
                    },
                })
        cls._transitions |= set((
                ('draft', 'calculated'),
                ('draft', 'cancelled'),
                ('calculated', 'draft'),
                ('calculated', 'done'),
                ('calculated', 'cancelled'),
                ('done', 'cancelled'),
                ('cancelled', 'draft'),
                ))

    @staticmethod
    def default_state():
        return 'draft'

    @staticmethod
    def default_support_type():
        return 'T'

    @staticmethod
    def default_type():
        return 'N'

    @staticmethod
    def default_company():
        return Transaction().context.get('company')

    @staticmethod
    def default_fiscalyear():
        FiscalYear = Pool().get('account.fiscalyear')
        return FiscalYear.find(
            Transaction().context.get('company'), exception=False)

    def get_rec_name(self, name):
        return '%s - %s/%s' % (self.company.rec_name,
            self.fiscalyear.name, self.period)

    def get_currency(self, name):
        return self.company.currency.id

    def get_filename(self, name):
        return 'aeat349-%s-%s.txt' % (
            self.fiscalyear_code, self.period)

    @fields.depends('fiscalyear')
    def on_change_with_fiscalyear_code(self):
        code = None
        if self.fiscalyear:
            code = self.fiscalyear.start_date.year
        return code

    @fields.depends('company')
    def on_change_with_company_vat(self):
        if self.company:
            tax_identifier = self.company.party.tax_identifier
            if tax_identifier and tax_identifier.code.startswith('ES'):
                return tax_identifier.code[2:]

    @classmethod
    def validate(cls, reports):
        for report in reports:
            report.check_euro()
            report.check_names()

    def check_euro(self):
        if self.currency.code != 'EUR':
            raise UserError(gettext('aeat_349.msg_invalid_currency',
                name=self.path,
                ))

    def check_names(self):
        """
        Checks that names are correct (not formed by only one string)
        """
        if self.state != 'done':
            return
        if not self.contact_name or len(self.contact_name.split()) < 2:
            raise UserError(gettext('aeat_349.msg_contact_name',
                name=self.path,
                ))

    @classmethod
    def get_totals(cls, reports, names):
        res = {}
        for name in ('operation_count', 'ammendment_count'):
            res[name] = dict.fromkeys([x.id for x in reports], 0)
        for name in ('operation_amount', 'ammendment_amount'):
            res[name] = dict.fromkeys([x.id for x in reports], _ZERO)
        for report in reports:
            res['operation_count'][report.id] = len(report.operations)
            res['operation_amount'][report.id] = (sum([
                        x.base for x in report.operations]) or Decimal('0.0'))
            res['ammendment_count'][report.id] = len(report.ammendments)
            res['ammendment_amount'][report.id] = (sum([
                        x.base for x in report.ammendments]) or Decimal('0.0'))
        for key in list(res.keys()):
            if key not in names:
                del res[key]
        return res

    @classmethod
    @ModelView.button
    @Workflow.transition('calculated')
    def calculate(cls, reports):
        pool = Pool()
        Data = pool.get('aeat.349.record')
        Operation = pool.get('aeat.349.report.operation')

        with Transaction().set_user(0):
            Operation.delete(Operation.search([
                ('report', 'in', [r.id for r in reports])]))

        for report in reports:
            fiscalyear = report.fiscalyear
            multiplier = 1
            period = report.period
            if 'T' in period:
                period = int(period[0]) - 1
                multiplier = 3
                start_month = period * multiplier + 1
            else:
                start_month = int(period) * multiplier

            end_month = start_month + multiplier

            to_create = {}
            for record in Data.search([
                    ('fiscalyear', '=', fiscalyear.id),
                    ('month', '>=', start_month),
                    ('month', '<', end_month)
                    ]):
                key = '%s-%s-%s' % (report.id, record.party_vat,
                    record.operation_key)

                if key in to_create:
                    to_create[key]['base'] += record.base
                    to_create[key]['records'][0][1].append(record.id)
                else:
                    to_create[key] = {
                        'base': record.base,
                        'party_vat': record.party_vat,
                        'party_name': record.party_name,
                        'operation_key': record.operation_key,
                        'report': report.id,
                        'records': [('add', [record.id])],
                    }
        with Transaction().set_user(0, set_context=True):
            Operation.create(list(to_create.values()))

        cls.write(reports, {
                'calculation_date': datetime.datetime.now(),
                })

    @classmethod
    @ModelView.button
    @Workflow.transition('done')
    def process(cls, reports):
        for report in reports:
            report.create_file()

    @classmethod
    @ModelView.button
    @Workflow.transition('cancelled')
    def cancel(cls, reports):
        pass

    @classmethod
    @ModelView.button
    @Workflow.transition('draft')
    def draft(cls, reports):
        pass

    def auto_sequence(self):
        pool = Pool()
        Report = pool.get('aeat.349.report')
        count = Report.search([
                ('state', '=', 'done'),
                ],
            order=[
                ('fiscalyear', 'DESC'),
                ('period', 'DESC'),
            ], count=True)
        return count + 1

    def create_file(self):
        records = []
        record = Record(aeat349.PRESENTER_HEADER_RECORD)
        record.fiscalyear = str(self.fiscalyear_code)
        record.nif = self.company_vat
        record.presenter_name = self.company.party.name
        record.support_type = self.support_type
        record.contact_phone = self.contact_phone
        record.contact_name = self.contact_name
        try:
            period = int(self.period)
        except ValueError:
            period = '0%s' % self.period[:1]
        record.declaration_number = int('349{}{}{:0>4}'.format(
            self.fiscalyear_code,
            period,
            self.auto_sequence()))
        record.complementary = '' if self.type == 'N' else self.type
        record.replacement = self.previous_number
        record.previous_declaration_number = self.previous_number
        record.period = self.period
        record.operation_count = len(self.operations)
        record.operation_amount = self.operation_amount or _ZERO
        record.ammendment_count = len(self.ammendments)
        record.ammendment_amount = self.ammendment_amount or _ZERO
        record.representative_nif = self.representative_vat
        records.append(record)
        for line in itertools.chain(self.operations, self.ammendments):
            record = line.get_record()
            record.fiscalyear = str(self.fiscalyear_code)
            record.nif = self.company_vat
            records.append(record)
        data = retrofix_write(records)
        data = remove_accents(data).upper()
        if isinstance(data, str):
            data = data.encode('iso-8859-1')
        self.file_ = self.__class__.file_.cast(data)
        self.save()


class Operation(ModelSQL, ModelView):
    """
    AEAT 349 Operation
    """
    __name__ = 'aeat.349.report.operation'
    _rec_name = 'party_name'

    company = fields.Function(fields.Many2One('company.company', 'Company'),
        'on_change_with_company', searcher='search_company')
    report = fields.Many2One('aeat.349.report', 'AEAT 349 Report',
        required=True)
    party_vat = fields.Char('VAT', size=17)
    party_name = fields.Char('Party Name', size=40)
    operation_key = fields.Selection(OPERATION_KEY, 'Operation key',
        required=True)
    base = fields.Numeric('Base Operation Amount', digits=(16, 2))
    records = fields.One2Many('aeat.349.record', 'operation',
        'AEAT 349 Records', readonly=True)

    @fields.depends('report', '_parent_report.company')
    def on_change_with_company(self, name=None):
        return self.report and self.report.company and self.report.company.id

    @classmethod
    def search_company(cls, name, clause):
        return [('report.%s' % name,) + tuple(clause[1:])]

    def get_record(self):
        record = Record(aeat349.OPERATOR_RECORD)
        record.party_vat = self.party_vat
        record.party_name = self.party_name
        record.operation_key = self.operation_key
        record.base = self.base or _ZERO
        return record


class Ammendment(ModelSQL, ModelView):
    """
    AEAT 349 Ammendment
    """
    __name__ = 'aeat.349.report.ammendment'

    company = fields.Many2One('company.company', 'Company', required=True)
    report = fields.Many2One('aeat.349.report', 'AEAT 349 Report')
    party_vat = fields.Char('VAT', size=17)
    party_name = fields.Char('Party Name', size=40)
    operation_key = fields.Selection(OPERATION_KEY, 'Operation key',
        required=True)
    ammendment_fiscalyear_code = fields.Integer('Ammendment Fiscal Year Code')
    ammendment_period = fields.Selection(PERIOD, 'Period', sort=False,
            required=True)
    base = fields.Numeric('Base Operation Amount', digits=(16, 2))
    original_base = fields.Numeric('Original Base', digits=(16, 2))

    @staticmethod
    def default_company():
        return Transaction().context.get('company')

    def get_record(self):
        record = Record(aeat349.AMMENDMENT_RECORD)
        record.party_vat = self.party_vat
        record.party_name = self.party_name
        record.operation_key = self.operation_key
        record.ammendment_fiscalyear = str(self.ammendment_fiscalyear_code)
        record.ammendment_period = self.ammendment_period
        record.base = self.base
        record.original_base = self.original_base
        return record
